pkg/nimble/rpble/nimble_rpble.c

```diff
@@ -0,0 +1,354 @@
+#include "nimble_netif_conn.h"
+#include "nimble_rpble.h"
+#include "nimble_scanner.h"
+#include "host/ble_gap.h"
+#include "nimble/nimble_port.h"
+
+#define ENABLE_DEBUG        0
+#include "debug.h"
+
+/* parent state */
+#define PARENT_NONE         -1
+#define PARENT_ROOT         -2
+#define SCORE_NONE          0U
+
+/* RPL meta data context fields */
+#define AD_SVC_DATA_LEN     23
+#define POS_INST_ID         2
+#define POS_DODAG_ID        3
+#define POS_VERSION         19
+#define POS_RANK            20
+#define POS_FREE_SLOTS      22
+
+/* keep the timing parameters for connections and advertisements */
+static struct ble_gap_adv_params _adv_params = { 0 };
+static struct ble_gap_conn_params _conn_params = { 0 };
+static uint32_t _conn_scan_to;      /* in ms */
+static uint32_t _conn_itvl_min;     /* in ms */
+static uint32_t _conn_itvl_max;     /* in ms */
+
+/* local RPL context */
+static nimble_rpble_ctx_t _local_rpl_ctx;
+static int _current_parent = PARENT_NONE;
+struct {
+    ble_addr_t addr;        /**< address of highest scored potential parent */
+    uint16_t score;         /* 0 := not used, larger is better! */
+} _psel = { { 0 }, SCORE_NONE };
+
+/* eval event used for periodical state updates */
+static uint32_t _eval_itvl;
+static struct ble_npl_callout _evt_eval;
+
+static nimble_netif_eventcb_t _eventcb = NULL;
+
+static uint16_t _psel_score(uint16_t rank, uint8_t free)
+{
+    return (UINT16_MAX - rank + free);
+}
+
+static void _children_accept(void)
+{
+    int res;
+    (void)res;
+
+    /* stop advertising while we are updating the context */
+    nimble_netif_accept_stop();
+
+    /* we only start advertising (accepting) if we do have an active parent and
+     * if we have resources for new connections */
+    if ((_current_parent == PARENT_NONE) ||
+        (nimble_netif_conn_count(NIMBLE_NETIF_UNUSED) == 0)) {
+        return;
+    }
+
+    /* compile RPL meta data into AD service data TLV field */
+    uint8_t sd[AD_SVC_DATA_LEN];
+    byteorder_htolebufs(sd, BLE_GATT_SVC_IPSS);
+    sd[POS_INST_ID] = _local_rpl_ctx.inst_id;
+    memcpy(&sd[POS_DODAG_ID], _local_rpl_ctx.dodag_id, 16);
+    sd[POS_VERSION] = _local_rpl_ctx.version;
+    byteorder_htolebufs(&sd[POS_RANK], _local_rpl_ctx.rank);
+    sd[POS_FREE_SLOTS] = (uint8_t)nimble_netif_conn_count(NIMBLE_NETIF_UNUSED);
+
+    /* generate the new advertising data */
+    bluetil_ad_t ad;
+    uint8_t buf[BLE_HS_ADV_MAX_SZ];
+    res = bluetil_ad_init_with_flags(&ad, buf, BLE_HS_ADV_MAX_SZ,
+                                     BLUETIL_AD_FLAGS_DEFAULT);
+    assert(res == BLUETIL_AD_OK);
+    res = bluetil_ad_add(&ad, BLE_GAP_AD_SERVICE_DATA, sd, sizeof(sd));
+    assert(res == BLUETIL_AD_OK);
+
+    /* start advertising this node */
+    res = nimble_netif_accept(ad.buf, ad.pos, &_adv_params);
+    assert(res == NIMBLE_NETIF_OK);
+}
+
+static void _on_scan_evt(uint8_t type,
+                         const ble_addr_t *addr, int8_t rssi,
```

Comment:
nice so we can hook this up to `netstats_nb_record()` and use it for MR-HOF :wink: